// @node-red-tools/node-red-contrib-mongodb/lib/core/collection/method.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var count_documents_1 = require("./methods/count-documents");
var create_index_1 = require("./methods/create-index");
var create_indexes_1 = require("./methods/create-indexes");
var delete_many_1 = require("./methods/delete-many");
var delete_one_1 = require("./methods/delete-one");
var distinct_1 = require("./methods/distinct");
var drop_index_1 = require("./methods/drop-index");
var drop_indexes_1 = require("./methods/drop-indexes");
var find_1 = require("./methods/find");
var find_one_1 = require("./methods/find-one");
var find_one_and_delete_1 = require("./methods/find-one-and-delete");
var find_one_and_replace_1 = require("./methods/find-one-and-replace");
var find_one_and_update_1 = require("./methods/find-one-and-update");
var insert_many_1 = require("./methods/insert-many");
var insert_one_1 = require("./methods/insert-one");
var map_reduce_1 = require("./methods/map-reduce");
var replace_one_1 = require("./methods/replace-one");
var update_many_1 = require("./methods/update-many");
var update_one_1 = require("./methods/update-one");
var METHODS = {
    countDocuments: {
        fn: count_documents_1.countDocuments,
        signature: count_documents_1.countDocumentsSignature,
    },
    createIndex: {
        fn: create_index_1.createIndex,
        signature: create_index_1.createIndexSignature,
    },
    createIndexes: {
        fn: create_indexes_1.createIndexes,
        signature: create_indexes_1.createIndexesSignature,
    },
    deleteMany: {
        fn: delete_many_1.deleteMany,
        signature: delete_many_1.deleteManySignature,
    },
    deleteOne: {
        fn: delete_one_1.deleteOne,
        signature: delete_one_1.deleteOneSignature,
    },
    distinct: {
        fn: distinct_1.distinct,
        signature: distinct_1.distinctSignature,
    },
    dropIndex: {
        fn: drop_index_1.dropIndex,
        signature: drop_index_1.dropIndexSignature,
    },
    dropIndexes: {
        fn: drop_indexes_1.dropIndexes,
        signature: drop_indexes_1.dropIndexesSignature,
    },
    find: {
        fn: find_1.find,
        signature: find_1.findSignature,
    },
    findOne: {
        fn: find_one_1.findOne,
        signature: find_one_1.findOneSignature,
    },
    findOneAndDelete: {
        fn: find_one_and_delete_1.findOneAndDelete,
        signature: find_one_and_delete_1.findOneAndDeleteSignature,
    },
    findOneAndReplace: {
        fn: find_one_and_replace_1.findOneAndReplace,
        signature: find_one_and_replace_1.findOneAndReplaceSignature,
    },
    findOneAndUpdate: {
        fn: find_one_and_update_1.findOneAndUpdate,
        signature: find_one_and_update_1.findOneAndUpdateSignature,
    },
    insertMany: {
        fn: insert_many_1.insertMany,
        signature: insert_many_1.insertManySignature,
    },
    insertOne: {
        fn: insert_one_1.insertOne,
        signature: insert_one_1.insertOneSignature,
    },
    mapReduce: {
        fn: map_reduce_1.mapReduce,
        signature: map_reduce_1.mapReduceSignature,
    },
    replaceOne: {
        fn: replace_one_1.replaceOne,
        signature: replace_one_1.replaceOneSignature,
    },
    updateMany: {
        fn: update_many_1.updateMany,
        signature: update_many_1.updateManySignature,
    },
    updateOne: {
        fn: update_one_1.updateOne,
        signature: update_one_1.updateOneSignature,
    },
};
function list() {
    return Object.keys(METHODS).map(function (name) {
        return {
            name: name,
            signature: METHODS[name].signature,
        };
    });
}
exports.list = list;
function resolve(name) {
    var method = METHODS[name];
    if (method) {
        return method.fn;
    }
    return undefined;
}
exports.resolve = resolve;
